kernel/ioapic/src/lib.rs

```diff
@@ -140,18 +135,18 @@ impl IoApic {
     ///
     /// # Arguments
     /// * `ioapic_irq`: the IRQ number that this interrupt will trigger on this IoApic.
-    /// * `lapic_id`: the id of the LocalApic that should handle this interrupt.
+    /// * `apic_id`: the ID of the Local APIC, i.e., the CPU, that should handle this interrupt.
     /// * `irq_vector`: the system-wide IRQ vector number,
     ///    which after remapping is from 0x20 to 0x2F 
     ///    (see [`interrupts::IRQ_BASE_OFFSET`](../interrupts/constant.IRQ_BASE_OFFSET.html)).
     ///    For example, 0x20 is the PIT timer, 0x21 is the PS2 keyboard, etc.
-    pub fn set_irq(&mut self, ioapic_irq: u8, lapic_id: u8, irq_vector: u8) {
+    pub fn set_irq(&mut self, ioapic_irq: u8, apic_id: ApicId, irq_vector: u8) {
         let low_index: u32 = 0x10 + (ioapic_irq as u32) * 2;
         let high_index: u32 = 0x10 + (ioapic_irq as u32) * 2 + 1;
 
         let mut high = self.read_reg(high_index);
         high &= !0xff000000;
-        high |= (lapic_id as u32) << 24;
+        high |= apic_id.value() << 24;
```

Comment:
I suggest checking that apic_id < 256

Reply:
ah, thanks for catching this. Checking that apic_id < 256 isn't actually the right solution, but rather I think it's necessary to use a bitshift of 24 for only xapic, and no bitshift for x2apic. I'll double-check.

Reply:
Turns out that you were indeed correct, that's a limitation of IOAPIC redirection tables. man, sometimes i hate x86. 
Anyway, i've fixed it now via an explicit check.